libsolidity/experimental/analysis/TypeClassRegistration.h

```diff
@@ -0,0 +1,58 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#pragma once
+
+#include <libsolidity/ast/ASTVisitor.h>
+#include <libsolidity/experimental/ast/TypeSystem.h>
+
+namespace solidity::langutil
+{
+class ErrorReporter;
+}
+
+namespace solidity::frontend::experimental
+{
+
+class Analysis;
+class TypeSystem;
+
+class TypeClassRegistration: public ASTConstVisitor
+{
+public:
+	struct Annotation
+	{
+		// Type classes.
+		std::optional<TypeClass> typeClass;
+	};
+	struct GlobalAnnotation
+	{
+	};
```

Comment:
Perhaps I should move more things from `TypeInference` here? `builtinClasses` and `builtinClassesByName` looks like good candidates.

I tried that already with `typeClassFunctions` but the problem is that I don't have the function types to put in it - those are only available after `TypeInference` processes function definitions. I could just gather `FunctionDefinitions` but `TypeInference` would have to iterate over them again and it did not seem like it would save us much code.